pin/set.go

```diff
@@ -0,0 +1,337 @@
+
+type refcount uint8
+
+func (r refcount) Bytes() []byte {
+	// refcount size can change in later versions; this may need
+	// encoding/binary
+	return []byte{byte(r)}
+}
+
+type sortByHash struct {
+	links []*merkledag.Link
+	data  []byte
+}
+
+func (s sortByHash) Len() int {
+	return len(s.links)
+}
+
+func (s sortByHash) Less(a, b int) bool {
+	return bytes.Compare(s.links[a].Hash, s.links[b].Hash) == -1
+}
+
+func (s sortByHash) Swap(a, b int) {
+	s.links[a], s.links[b] = s.links[b], s.links[a]
+	if len(s.data) != 0 {
+		const n = int(unsafe.Sizeof(refcount(0)))
+		tmp := make([]byte, n)
+		copy(tmp, s.data[a:a+n])
+		copy(s.data[a:a+n], s.data[b:b+n])
+		copy(s.data[b:b+n], tmp)
+	}
+}
+
+func storeItems(ctx context.Context, dag merkledag.DAGService, estimatedLen uint64, iter itemIterator, internalKeys keyObserver) (*merkledag.Node, error) {
+	seed, err := randomSeed()
+	if err != nil {
+		return nil, err
+	}
+	n := &merkledag.Node{
+		Links: make([]*merkledag.Link, 0, defaultFanout+maxItems),
+	}
+	for i := 0; i < defaultFanout; i++ {
+		n.Links = append(n.Links, &merkledag.Link{Hash: emptyKey.ToMultihash()})
+	}
+	internalKeys(emptyKey)
+	hdr := &pb.Set{
+		Version: proto.Uint32(1),
+		Fanout:  proto.Uint32(defaultFanout),
+		Seed:    proto.Uint32(seed),
+	}
+	if err := writeHdr(n, hdr); err != nil {
+		return nil, err
+	}
+	hdrLen := len(n.Data)
+
+	if estimatedLen < maxItems {
+		// it'll probably fit
+		for i := 0; i < maxItems; i++ {
+			k, data, ok := iter()
+			if !ok {
+				// all done
+				break
+			}
+			n.Links = append(n.Links, &merkledag.Link{Hash: k.ToMultihash()})
+			n.Data = append(n.Data, data...)
+		}
+		// sort by hash, also swap item Data
+		s := sortByHash{
+			links: n.Links[defaultFanout:],
+			data:  n.Data[hdrLen:],
+		}
+		sort.Stable(s)
+	}
+
+	// wasteful but simple
+	type item struct {
+		k    util.Key
+		data []byte
+	}
+	hashed := make(map[uint32][]item)
+	for {
+		k, data, ok := iter()
+		if !ok {
+			break
+		}
+		h := hash(seed, k)
+		hashed[h] = append(hashed[h], item{k, data})
+	}
+	for h, items := range hashed {
+		childIter := func() (k util.Key, data []byte, ok bool) {
+			if len(items) == 0 {
+				return "", nil, false
+			}
+			first := items[0]
+			items = items[1:]
+			return first.k, first.data, true
+		}
+		child, err := storeItems(ctx, dag, uint64(len(items)), childIter, internalKeys)
+		if err != nil {
+			return nil, err
+		}
+		size, err := child.Size()
+		if err != nil {
+			return nil, err
+		}
+		childKey, err := dag.Add(child)
+		if err != nil {
+			return nil, err
+		}
+		internalKeys(childKey)
+		l := &merkledag.Link{
+			Name: "",
+			Hash: childKey.ToMultihash(),
+			Size: size,
+			Node: child,
+		}
+		n.Links[int(h%defaultFanout)] = l
+	}
+	return n, nil
+}
+
+func readHdr(n *merkledag.Node) (*pb.Set, []byte, error) {
+	hdrLenRaw, consumed := binary.Uvarint(n.Data)
+	if consumed <= 0 {
+		return nil, nil, errors.New("invalid Set header length")
+	}
+	buf := n.Data[consumed:]
```

Comment:
That would be a bug in encoding/binary, I believe. And it's returning just i+1 where i is the index from a range, and it has a special check for consuming max 10 bytes; it's really not gonna suddenly be something crazy.